### PR TITLE
fix: restrict DM agent retrieval to shared streams

### PR DESCRIPTION
**Pull request by kristofferremback**
## Problem

Ariadnes DM retrieval scope used the union of every participants visible streams. In practice that let a DM between Kristoffer and Agnes pull workspace memory from Kristoffer-only private streams, including the Kristoffer/Pierre DM, Kristoffers private scratchpads, and his private system stream.

## Solution

Restrict DM retrieval to the intersection of streams visible to all DM participants.

### How it works

1. DM invocations now produce a `user_intersection` access spec instead of `user_union`.
2. `SearchRepository.getAccessibleStreamsForAgent()` resolves each participants visible streams and keeps only the shared set.
3. `WorkspaceAgent` uses that shared stream set for `workspace_research`, message search, memo search, and attachment search in DMs.

### Key design decisions

**1. Fix the shared access layer, not individual tools**

The leak came from the common `accessibleStreamIds` calculation. Fixing the access spec and search repository closes the hole for `workspace_research` and the other workspace tools without duplicating permission logic in each tool.

**2. Preserve existing behavior outside DMs**

Scratchpads still use invoking-user access, and private channels still use public-plus-current-stream access. The change is intentionally limited to DM retrieval semantics.

**3. Add a regression test at the exact boundary that failed**

The new integration test exercises `computeAgentAccessSpec()` plus `SearchRepository.getAccessibleStreamsForAgent()` against shared channels, private scratchpads, and unrelated DMs so the privacy boundary is enforced where the bug originated.

## New files

| File | Purpose |
| --- | --- |
| `apps/backend/tests/integration/agent-access-scope.test.ts` | Regression test proving DM agent retrieval only sees streams shared by all participants. |

## Modified files

| File | Change |
| --- | --- |
| `apps/backend/src/features/agents/researcher/access-spec.ts` | Replaced DM `user_union` access with `user_intersection` and updated documentation/debug output. |
| `apps/backend/src/features/agents/researcher/researcher.ts` | Updated DM workspace-research setup to use intersection semantics. |
| `apps/backend/src/features/search/repository.ts` | Changed DM agent stream resolution from union to intersection across participant-visible streams. |

## Deleted files

None.

## Test plan

- [x] `bun scripts/test-silent.ts backend-integration tests/integration/agent-access-scope.test.ts`
- [x] `bun run --cwd apps/backend typecheck`
- [x] Pre-commit hooks ran backend/control-plane/frontend lint plus repo typechecks successfully
- [ ] Deploy the backend and verify Ariadne in a DM no longer cites private streams that are not shared by both participants

---

🤖 _PR by [Claude Code](https://claude.com/claude-code)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Direct message access tightened: agents now see only DMs shared by all participants (intersection); invalid DM participant counts are rejected.

* **Refactor**
  * Simplified DM handling across agent context and search flows: removed per-DM participant overrides so access is based on the computed spec.

* **Tests**
  * Added integration tests for agent access scope, including DM scenarios and attachment-search restrictions.

* **Chores**
  * Added a constant to enforce the required DM participant count.

### CODE REVIEW

File: apps/backend/src/features/agents/researcher/researcher.ts [hunk 1]
```diff
@@ -187,9 +187,9 @@ export class WorkspaceAgent {
         invokingUserId,
       })
 
-      // For DMs, we need to pass participant IDs
+      // For DMs, retrieval must stay within streams all participants can access.
       const effectiveAccessSpec: AgentAccessSpec =
-        stream.type === "dm" && dmParticipantIds ? { type: "user_union", userIds: dmParticipantIds } : accessSpec
+        stream.type === "dm" && dmParticipantIds ? { type: "user_intersection", userIds: dmParticipantIds } : accessSpec
```
[the remark this answers is not in the document]
Comment (kristofferremback) [reply apps/backend/src/features/agents/researcher/researcher.ts]: Fixed in `b9629d42`.

- Removed the caller-supplied `dmParticipantIds` override entirely, so `WorkspaceAgent` now uses the DB-derived `computeAgentAccessSpec()` result as the single source of truth for DM participant scope.
- Added fail-closed guards for malformed DM intersection specs: DMs now require exactly 2 members when computing the access spec, and `user_intersection` rejects fewer than two users if constructed manually.
- Tightened the regression test to assert the exact shared stream set and added a malformed-intersection case.

🤖 _Response by [Claude Code](https://claude.com/claude-code)_

**Comment by kristofferremback** [comment]
### Code review
**Confidence: 5/7** — Good
**Review models:** Orchestrator: claude-sonnet-4-6 | Reviewers: sonnet x3

Found 2 issues:

1. `repository.ts:456-458` — Dead code (CLAUDE.md [INV-36/INV-38, Maintainability and Scope Control]: "Do not add speculative features, configuration, or comments for imagined requirements" / "Delete dead code immediately"): the `if (allResults.length === 0) { return [] }` guard is unreachable. The preceding throw at line 441 (`spec.userIds.length < 2`) guarantees `Promise.all(spec.userIds.map(...))` always produces ≥2 results.
   https://github.com/kristofferremback/threa/blob/b9629d42861d098ea1dd34b89ae4b0cab71a835c/apps/backend/src/features/search/repository.ts#L454-L460

2. `repository.ts:439-463` — N parallel queries + JS set intersection where a single SQL `INTERSECT` (or CTE) would be one round-trip (CLAUDE.md [INV-56, Data Model and Persistence Safety]: "Prefer set-based/batch DB operations over per-row loops"). `getAccessibleStreamsWithMembers` anchors on a single `userId` so multi-user intersection cannot be expressed as one call, forcing N queries + application-layer set arithmetic. For the current DM=2 case this is bounded, but the repository guard is `< 2` while `access-spec.ts` enforces `!== 2`, meaning the repository silently accepts N>2 user specs and would issue N queries with no upper bound.
   https://github.com/kristofferremback/threa/blob/b9629d42861d098ea1dd34b89ae4b0cab71a835c/apps/backend/src/features/search/repository.ts#L437-L465

---
<details><summary>📐 Plan Adherence [CLEAN]</summary>

Diff matches PR description. `computeAgentAccessSpec` is the single access-spec source for DMs; `dmParticipantIds` override has been removed from the caller chain. No missing corresponding changes.
</details>
<details><summary>🔍 Bugs [CLEAN]</summary>

No logic errors, null hazards, or race conditions. Intersection algorithm correct. DB migration `20260217195856_dm_one_to_one_constraints.sql` enforces the 2-member DM invariant at the data layer, making the `members.length !== 2` guard in `access-spec.ts` sound.
</details>
<details><summary>📋 CLAUDE.md Compliance [1 concern]</summary>

- `repository.ts:456-458` — INV-36/INV-38 dead code (see issue #1)
</details>
<details><summary>🏗️ Design [1 concern]</summary>

- `repository.ts:439-463` — N queries + JS set intersection instead of a single SQL `INTERSECT`. The `!== 2` producer guard vs `< 2` repository guard also leaves no single source of truth for what a valid `user_intersection` spec contains (flagged in previous review, still present).
</details>
<details><summary>🔒 Security [CLEAN]</summary>

No exploitable vulnerabilities. Union→intersection is security-positive. No new data-flow paths bypass workspace or stream membership checks.
</details>

🤖 Generated with unified-review automation
<sub>If this review was useful, react with 👍. Otherwise, react with 👎.</sub>
